include/pybind11/pytypes.h

```diff
@@ -1059,6 +1059,34 @@ inline str::str(const bytes& b) {
     m_ptr = obj.release().ptr();
 }
 
+/// \addtogroup pytypes
+/// @{
+class bytearray : public object {
+public:
+    PYBIND11_OBJECT_CVT(bytearray, object, PyByteArray_Check, PyByteArray_FromObject)
+
+    bytearray(const char *c, size_t n)
+        : object(PyByteArray_FromStringAndSize(c, (ssize_t) n), stolen_t{}) {
+        if (!m_ptr) pybind11_fail("Could not allocate bytearray object!");
+    }
+
+    bytearray()
+        : bytearray("", 0) {}
+
+    explicit bytearray(const std::string &s) : bytearray(s.data(), s.size()) { }
+
+    size_t size() const { return static_cast<size_t>(PyByteArray_Size(m_ptr)); }
+
+    explicit operator std::string() const {
+        char *buffer = PyByteArray_AS_STRING(m_ptr);
```

Comment:
One last time. Are we sure we don't want to check for NULL and throw?

Reply:
ok I will add that otherwise the code may become risky.

Reply:
As I can see [here](https://github.com/python/cpython/blob/a6192635f1e62af2bb8a435487ebb51800edd671/Objects/bytearrayobject.c#L159), `PyByteArray_AsString` does not check for `NULL` at all, except for some `assert`s which you said maybe disabled at runtime. So, we have check for `nullptr` on our side.
If I find `m_ptr` is `NULL`, since I can't throw `error_already_set`, which exception should I throw?

Reply:
Right... It just does this:

https://github.com/python/cpython/blob/63298930fb531ba2bb4f23bc3b915dbf1e17e9e1/Include/cpython/bytearrayobject.h#L15-L17

It still might dereference `self`. I believe that in these cases pybind11 uses `pybind11_fail(what_str)`.

Reply:
Oh right

Reply:
Done.

Reply:
Actually, I was about to argue that this was fine. You shouldn't be using a `NULL` `py::object` (and you can check for it with `is_none()`), and the constructors don't allow to construct one, so the only way of getting a `NULL` `bytearray` is through `reinterpret_borrow`/`reinterpret_steal`, which is an "advanced" operation anyway?

Reply:
Are you arguing that we should revert?

Reply:
I was wondering if this was necessary, yes. But unless you strongly agree, I'm not arguing to revert, since this thing has been going on for long enough.

Reply:
Isn't it better to keep the check anyways for future if someday the invariant that it will never be NULL is not true anymore?

Reply:
Myeah, maybe, but I think there's lots of other places where non-`NULL` is already assumed.

And then `PyByteArray_Size()` isn't much safer either?

https://github.com/python/cpython/blob/3cf0833f42ebde24f6435b838785ca4f946b988f/Objects/bytearrayobject.c#L149-L156

Reply:
But yes, it was just a suggestion. If you or @bstaletic aren't convinced, let's just go for this.

Reply:
> And then PyByteArray_Size() isn't much safer either?

Oh... in that case either we should check for NULL there as well, or not check for NULL in either place. @YannickJadoul is right when he says

> the only way of getting a NULL bytearray is through reinterpret_borrow/reinterpret_steal, which is an "advanced" operation anyway?

So I'm not against dropping the NULL check, but I would like to see consistent behaviour with respect to `m_ptr == NULL`.

Reply:
Ok, to keep it consistent with the other places, I will remove the NULL check.

Reply:
Should I go ahead and remove?

Reply:
Go ahead.

Reply:
Yeah, to be fair, pybind11 is already not consistent I think. Sometimes safe methods are used, but ... yeah, I don't know :-) Nothing to be fixed in this PR, anyway.

Reply:
If users come with a good complaint about why we should check for `NULL` instead of them, we can still add a check.

Reply:
Done.